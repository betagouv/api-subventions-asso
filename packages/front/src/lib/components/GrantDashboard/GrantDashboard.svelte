<script lang="ts">
    import DataNotFound from "../DataNotFound.svelte";
    import ErrorAlert from "../ErrorAlert.svelte";
    import { GrantDashboardController } from "./GrantDashboard.controller";
    import ApplicationRow from "./ApplicationRow/ApplicationRow.svelte";
    import PaymentRow from "./PaymentRow/PaymentRow.svelte";
    import GrantsStatistique from "$lib/components/GrantDashboard/GrantsStatistique/GrantsStatistique.svelte";
    import Spinner from "$lib/components/Spinner.svelte";
    import Button from "$lib/dsfr/Button.svelte";
    import Select from "$lib/dsfr/Select.svelte";
    import Table from "$lib/dsfr/Table.svelte";
    import TableRow from "$lib/dsfr/TableRow.svelte";

    export let structureId;

    const tableId = "grant-dashboard";

    const ctrl = new GrantDashboardController(structureId);
    const {
        grantPromise,
        grants,
        headers,
        rows,
        selectedExerciseIndex,
        selectedExercise,
        exerciseOptions,
        isExtractLoading,
        selectedGrants,
    } = ctrl;
</script>

{#await grantPromise}
    <Spinner description="Chargement des demandes de subventions en cours ..." />
{:then _null}
    <div class="fr-grid-row fr-pt-4w fr-pb-2w flex space-between">
        <h2>Subventions</h2>
        <div class="baseline">
            {#if $grants && $grants.length}
                {#if $isExtractLoading}
                    <Button icon="refresh-line" trackingDisable={true} iconPosition="right" disabled={true}>
                        Chargement
                    </Button>
                {:else}
                    <Button
                        on:click={() => ctrl.download()}
                        icon="download-line"
                        trackingDisable={true}
                        iconPosition="right">
                        Télécharger les données
                    </Button>
                {/if}
            {/if}
        </div>
    </div>
    <div class="fr-grid-row flex space-between">
        <div class="fr-col-3">
            {#if $exerciseOptions?.length}
                <Select
                    on:change={event => ctrl.selectExercise(event.detail)}
                    selected={$selectedExerciseIndex}
                    options={$exerciseOptions} />
            {/if}
        </div>
        <div class="align-bottom">
            <a
                on:click={ctrl.clickProviderLink}
                class="fr-link"
                href={ctrl.providerBlogUrl}
                target="_blank"
                rel="noopener external">
                Quelles données retrouver dans Data.Subvention ?
            </a>
        </div>
    </div>
    <div class="fr-mt-6w compact-columns">
        {#if $selectedGrants?.length}
            <div>
                <GrantsStatistique grants={$grants} year={$selectedExercise} />
            </div>
            <div class="negative-margin">
                <Table
                    id={tableId}
                    on:sort={event => ctrl.sortTable(event.detail)}
                    title="Tableau de subventions et leurs versements"
                    hideTitle={true}
                    size="md"
                    sortable={true}
                    scrollable={false}
                    bordered={false}
                    {headers}>
                    {#each $rows as row, rowIndex}
                        <TableRow id={tableId} index={rowIndex} openModal={true}>
                            <ApplicationRow
                                on:click={() => ctrl.onApplicationClick(rowIndex)}
                                cells={row.applicationCells} />
                            <PaymentRow
                                on:click={() => ctrl.onPaymentClick(rowIndex)}
                                cells={row.paymentsCells}
                                granted={row.granted} />
                        </TableRow>
                    {/each}
                </Table>
            </div>
        {:else}
            <DataNotFound content={ctrl.notFoundMessage} />
        {/if}
    </div>
{:catch error}
    {#if error.request && error.request.status == 404}
        <DataNotFound />
    {:else}
        <ErrorAlert message={error.message} />
    {/if}
{/await}

<style>
    .baseline {
        align-self: baseline;
    }

    /* remove padding from fr-table__container and margin from fr-table */
    .negative-margin {
        margin-top: -2rem;
        margin-bottom: -2.5rem;
    }
</style>
