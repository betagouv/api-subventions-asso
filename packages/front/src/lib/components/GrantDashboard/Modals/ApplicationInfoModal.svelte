<script>
    import { data } from "$lib/store/modal.store";
    import { capitalizeFirstLetter } from "$lib/helpers/stringHelper";
    import { numberToEuro } from "$lib/helpers/dataHelper.js";
</script>

{#if $data.application}
    <section>
        <h4 class="fr-icon-arrow-right-line">Informations collectées</h4>
        <p class="fr-text--lead">
            {#if $data.application.montants.accorde}
                <span class="fr-text--bold">{numberToEuro($data.application.montants.accorde)}</span>
                ont été accordés{#if $data.application.montants.demande}
                    sur <span class="fr-text--bold">{numberToEuro($data.application.montants.demande)}</span>
                    demandés{/if}.
            {:else if $data.application.montants.demande}
                <span class="fr-text--bold">{numberToEuro($data.application.montants.demande)}</span>
                ont été demandés.
            {/if}
        </p>
        {#if $data.application.date_depot}
            <p class="fr-text--lg">
                <span class="fr-text--bold">Date de dépôt de la demande :</span>
                {$data.application.date_depot}
            </p>{/if}
        {#if $data.application.date_decision}
            <p class="fr-text--lg">
                <span class="fr-text--bold">Date de décision du service instructeur :</span>
                {$data.application.date_decision}
            </p>{/if}
    </section>
    {#each $data.application?.actions_proposee || [] as action, index (index)}
        <section>
            <h4 class="fr-icon-arrow-right-line">{capitalizeFirstLetter(action.intitule)}</h4>
            {#each action.objectifs?.split("\n") || [] as line, index (index)}
                {#if line.length}
                    <p>{line}</p>
                {/if}
            {/each}
        </section>
    {/each}
{/if}

<style>
    section h4 {
        color: var(--text-active-blue-france);
        margin-bottom: 1.33em;
    }

    section {
        margin-bottom: 3rem;
    }

    section:last-child {
        margin-bottom: 0;
    }

    section h4:is([class*=" fr-fi-"], h4[class*=" fr-icon-"], h4[class^="fr-fi-"], h4[class^="fr-icon-"])::before {
        margin-right: 0.5rem;
    }

    /* uncomment when showing establishment
    .bigger-link {
        position: relative;
    }

    .bigger-link > a::before {
        position: absolute;
        width: 100%;
        height: 100%;
        content: "";
    }*/
</style>
