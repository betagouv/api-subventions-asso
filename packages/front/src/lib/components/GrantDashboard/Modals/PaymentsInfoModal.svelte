<script>
    import Table from "$lib/dsfr/Table.svelte";
    import TableRow from "$lib/dsfr/TableRow.svelte";

    import { data } from "$lib/store/modal.store";
    import { numberToEuro, valueOrHyphen } from "$lib/helpers/dataHelper.js";
    import { dateToDDMMYYYY } from "$lib/helpers/dateHelper.js";

    const headers = ["Montant", "Domaine fonctionnel", "Activité", "Centre financier", "Date", "Programme"];
    const tableId = "payments-modal";
</script>

{#if $data.payments}
    <Table id={tableId} {headers}>
        {#each $data.payments as payment, index}
            <TableRow id={tableId} {index} title="Détail des versements" hideTitle={true}>
                <td class="primary">{valueOrHyphen(numberToEuro(payment.amount))}</td>
                <td>{valueOrHyphen(payment.domaineFonctionnel)}</td>
                <td>{valueOrHyphen(payment.activitee)}</td>
                <td>{valueOrHyphen(payment.centreFinancier)}</td>
                <td>{valueOrHyphen(dateToDDMMYYYY(new Date(payment.dateOperation)))}</td>
                <td>{payment.programme ? `${payment.programme} - ${valueOrHyphen(payment.libelleProgramme)}` : "-"}</td>
            </TableRow>
        {/each}
    </Table>
{/if}
