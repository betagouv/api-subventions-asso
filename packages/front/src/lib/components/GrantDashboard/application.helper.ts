import { ApplicationStatus } from "dto";
import type { FlatApplication } from "$lib/resources/@types/FlattenGrant";
import type { TableCell } from "$lib/dsfr/TableCell.types";
import { numberToEuro } from "$lib/helpers/dataHelper";
import { withTwoDigitYear } from "$lib/helpers/dateHelper";
import { capitalizeFirstLetter } from "$lib/helpers/stringHelper";
import { mapSiretPostCodeStore } from "$lib/store/association.store";

export const getApplicationCells = (application: FlatApplication | null, accepted: boolean): TableCell[] | null => {
    if (!application) return null;
    else {
        const grantedAmount = numberToEuro(application.montants?.accorde);

        const requestedCell = application.montants?.demande
            ? {
                  title: numberToEuro(application.montants?.demande),
                  desc: application.creer_le
                      ? `DemandÃ© le ${withTwoDigitYear(new Date(application.creer_le))}`
                      : undefined,
              }
            : { desc: null };

        return [
            { desc: mapSiretPostCodeStore.value.get(application.siret?.toString() as string) },
            { desc: application.service_instructeur },
            { desc: application.dispositif },
            { desc: getProjectName(application) },
            requestedCell,
            {
                badge: accepted ? { type: "success", label: grantedAmount } : { status: application.statut_label },
            },
        ];
    }
};

export function getApplicationDashboardData(application: FlatApplication | null) {
    if (!application) return application;
    return {
        montantAccorde: application.montants?.accorde,
        montantDemande: application.montants?.demande,
        date_demande: application.creer_le ? new Date(application.creer_le) : undefined,
        postal_code: mapSiretPostCodeStore.value.get(application.siret?.toString() as string),
        service_instructeur: application.service_instructeur,
        dispositif: application.dispositif,
        nomProjet: getProjectName(application),
        statut_label: application.statut_label,
    };
}

export const isGranted = (application: FlatApplication | null) => {
    return application?.statut_label === ApplicationStatus.GRANTED;
};

export const getProjectName = application => {
    if (!application) return undefined;
    if (!application.actions_proposee || !application.actions_proposee.length) return;

    let names = application.actions_proposee
        .sort((actionA, actionB) => actionA.rang - actionB.rang)
        .map(action => `${capitalizeFirstLetter(action.intitule)}.`.replace(/\.\./g, "."));

    names = [...new Set(names)].join(" - ");

    return names;
};
