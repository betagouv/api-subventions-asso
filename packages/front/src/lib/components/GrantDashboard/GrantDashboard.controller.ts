import type { StructureIdentifierDto } from "dto";
import ApplicationInfoModal from "./Modals/ApplicationInfoModal.svelte";
import PaymentsInfoModal from "./Modals/PaymentsInfoModal.svelte";
import { getApplicationCells, getApplicationDashboardData, isGranted } from "./application.helper";
import { getPaymentDashboardData, getPaymentsCells } from "./payments.helper";
import type { FlatGrant, OnlyApplication } from "$lib/resources/@types/FlattenGrant";
import { isSiret } from "$lib/helpers/identifierHelper";
import associationPort from "$lib/resources/associations/association.port";
import establishmentPort from "$lib/resources/establishments/establishment.port";
import { returnInfinitePromise } from "$lib/helpers/promiseHelper";
import Store, { derived, ReadStore } from "$lib/core/Store";
import trackerService from "$lib/services/tracker.service";
import establishmentService from "$lib/resources/establishments/establishment.service";
import associationService from "$lib/resources/associations/association.service";
import documentHelper from "$lib/helpers/document.helper";
import { PUBLIC_PROVIDER_BLOG_URL } from "$env/static/public";
import { data, modal } from "$lib/store/modal.store";
import type { SortableRow } from "$lib/components/GrantDashboard/@types/DashboardGrant";
import { grantCompareFn } from "$lib/components/GrantDashboard/sort.helper";

export class GrantDashboardController {
    public identifier: StructureIdentifierDto;
    public grantPromise: Promise<FlatGrant[]> = returnInfinitePromise();
    public grants: Store<FlatGrant[] | undefined> = new Store(undefined);
    public grantsByExercise: Record<string, FlatGrant[]> = {};
    public selectedExerciseIndex: Store<number | undefined> = new Store(undefined);

    public selectedGrants: Store<FlatGrant[] | null> = new Store(null);
    public selectedExercise: ReadStore<string | null> = new ReadStore(null);

    // final rows displayed in view: can be updated with exercise filter and in a future with other filters
    public rows: Store<SortableRow[]> = new Store([]);
    public isExtractLoading: Store<boolean> = new Store(false);
    public exerciseOptions: Store<string[] | undefined> = new Store(undefined);
    public headers: string[];
    private columnsSortOrder: number[];

    constructor(identifier: StructureIdentifierDto) {
        this.identifier = identifier;

        // if you change this please update applicationCellsLength and paymentsCellsLength
        // TODO create this from a applicationHeader + paymentHeader to simplify above comment THIS
        this.headers = [
            "Code postal",
            "Instructeur",
            "Dispositif",
            "Action",
            "Demandé",
            "Statut",
            "Versé",
            "Programme",
        ];
        this.columnsSortOrder = Array(this.headers.length).fill(1);

        this.initStores();

        this.grantPromise = isSiret(identifier)
            ? establishmentPort.getGrants(identifier)
            : associationPort.getGrants(identifier);

        this.grantPromise.then(grants => this.processGrants(grants));
    }

    private initStores() {
        this.selectedGrants = new Store(null);
        this.selectedExercise = derived(this.selectedExerciseIndex, index => {
            if (index == null || !this.exerciseOptions.value) return null;
            return this.exerciseOptions.value[index];
        });
        this.selectedExercise.subscribe(exercise => {
            if (exercise == null) return;
            this.selectedGrants.set(this.grantsByExercise[exercise]);
        });
        this.selectedGrants.subscribe(grants => this.updateRows(grants));
    }

    /*
    initial processing of grants to split them by exercise and fill available exercises for select
     */
    private processGrants(grants: FlatGrant[]) {
        this.grants.set(grants);
        this.grantsByExercise = this.splitGrantsByExercise(this.grants.value as FlatGrant[]);

        this.exerciseOptions.set(Object.keys(this.grantsByExercise));
        this.selectedExerciseIndex.set((this.exerciseOptions.value as string[]).length - 1);
    }

    get providerBlogUrl() {
        return PUBLIC_PROVIDER_BLOG_URL;
    }

    /*
    message for no subvention at all
     */
    get notFoundMessage() {
        const defaultContent = "Nous sommes désolés, nous n'avons trouvé aucune donnée pour cet établissement";
        if (this.exerciseOptions.value?.length) return `${defaultContent} sur l'année ${this.selectedExercise.value}`;
        else return defaultContent;
    }

    private updateRows(grants) {
        if (!grants) return;
        this.rows.set(
            grants.map(grant => {
                const granted = isGranted(grant.application);
                return {
                    applicationCells: getApplicationCells(grant.application, granted),
                    paymentsCells: getPaymentsCells(grant.payments),
                    granted,
                    application: getApplicationDashboardData(grant.application),
                    payment: getPaymentDashboardData(grant.payments),
                };
            }),
        );
    }

    public sortTable(index: number) {
        // change order between ASC and DESC
        this.columnsSortOrder[index] *= -1;
        this.rows.update(rows => rows.sort((a, b) => grantCompareFn[index](a, b, this.columnsSortOrder[index])));
    }

    public selectExercise(index) {
        this.selectedExerciseIndex.set(index);
    }

    public clickProviderLink() {
        trackerService.trackEvent("association-etablissement.dashboard.display-provider-modal");
    }

    // Grants are expected to be ordered from API
    private splitGrantsByExercise(grants: FlatGrant[]) {
        return grants.reduce((byExercise, grant) => {
            let exercise;
            if (grant.payments?.length) exercise = String(new Date(grant.payments[0].dateOperation).getFullYear());
            else exercise = String((grant as OnlyApplication).application.annee_demande);
            if (!byExercise[exercise]) byExercise[exercise] = [grant];
            else byExercise[exercise].push(grant);
            return byExercise;
        }, {} as Record<string, FlatGrant[]>);
    }

    public async download() {
        if (this.isExtractLoading.value) return;
        trackerService.buttonClickEvent("association-etablissement.dashbord.download-csv", this.identifier); // tracking
        this.isExtractLoading.set(true);
        const resourceService = isSiret(this.identifier) ? establishmentService : associationService;
        await resourceService.getGrantExtract(this.identifier).then(({ blob, filename }) => {
            documentHelper.download(blob, filename);
        });
        this.isExtractLoading.set(false);
    }

    public onPaymentClick(index) {
        if (!this.rows.value[index].paymentsCells) return;
        trackerService.buttonClickEvent("association-etablissement.dashbord.payment.more_information");
        data.set({ payments: (this.selectedGrants.value as FlatGrant[])[index].payments });
        modal.set(PaymentsInfoModal);
    }

    public onApplicationClick(index) {
        if (!this.rows.value[index].applicationCells) return;
        trackerService.buttonClickEvent("association-etablissement.dashbord.subvention.more_information");
        data.set({ application: (this.selectedGrants.value as FlatGrant[])[index].application });
        modal.set(ApplicationInfoModal);
    }
}
