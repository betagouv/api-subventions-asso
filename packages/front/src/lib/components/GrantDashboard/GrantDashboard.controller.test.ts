import { GrantDashboardController } from "./GrantDashboard.controller";
import { getPaymentDashboardData, getPaymentsCells } from "./payments.helper";
import { getApplicationCells, getApplicationDashboardData, isGranted } from "./application.helper";
import type { FlatGrant } from "$lib/resources/@types/FlattenGrant";
import { isSiret } from "$lib/helpers/identifierHelper";
import establishmentService from "$lib/resources/establishments/establishment.service";
import establishmentPort from "$lib/resources/establishments/establishment.port";
import associationService from "$lib/resources/associations/association.service";
import associationPort from "$lib/resources/associations/association.port";
import trackerService from "$lib/services/tracker.service";
import documentHelper from "$lib/helpers/document.helper";
import PaymentsInfoModal from "$lib/components/GrantDashboard/Modals/PaymentsInfoModal.svelte";
import ApplicationInfoModal from "$lib/components/GrantDashboard/Modals/ApplicationInfoModal.svelte";
import { data, modal } from "$lib/store/modal.store";
import type {
    DashboardApplication,
    DashboardPayment,
    SortableRow,
} from "$lib/components/GrantDashboard/@types/DashboardGrant";
import { grantCompareFn } from "$lib/components/GrantDashboard/sort.helper";
import type { TableCell } from "$lib/dsfr/TableCell.types";
import * as Store from "$lib/core/Store";
import type { ReadStore } from "$lib/core/Store";

vi.mock("$lib/helpers/identifierHelper");
vi.mock("$lib/resources/establishments/establishment.service");
vi.mock("$lib/resources/establishments/establishment.port");
vi.mock("$lib/resources/associations/association.service");
vi.mock("$lib/resources/associations/association.port");
vi.mock("$lib/helpers/providerValueHelper");
vi.mock("$lib/services/tracker.service");
vi.mock("$lib/helpers/document.helper");
vi.mock("$lib/store/modal.store");
vi.mock("./application.helper");
vi.mock("./payments.helper");
vi.mock("$lib/components/GrantDashboard/sort.helper", () => ({
    grantCompareFn: [null, vi.fn((_a, _b, _sortOrder) => 0)],
}));

describe("GrantDashboard Controller", () => {
    const IDENTIFIER = "000000001";
    const INDEX = 1;

    const GRANT_FROM_2024 = () => ({
        application: { annee_demande: 2024 },
    });

    const GRANT_FROM_2023 = () => ({
        application: { annee_demande: 2022 },
        payments: [{ dateOperation: "2023-01-13" }, { dateOperation: "2023-04-21" }],
    });

    const GRANT_FROM_2021 = () => ({ payments: [{ dateOperation: "2021-08-14" }] });

    const FLAT_GRANTS = () => [GRANT_FROM_2021(), GRANT_FROM_2023(), GRANT_FROM_2024()] as FlatGrant[];

    let CTRL: GrantDashboardController;

    beforeAll(() => {
        vi.mocked(associationPort.getGrants).mockResolvedValue(FLAT_GRANTS());
        vi.mocked(establishmentPort.getGrants).mockResolvedValue(FLAT_GRANTS());
    });

    describe("constructor", () => {
        beforeAll(() => {
            vi.mocked(isSiret).mockReturnValue(true);
        });

        beforeEach(() => {
            CTRL = new GrantDashboardController(IDENTIFIER);
            // @ts-expect-error: mock
            vi.spyOn(CTRL, "processGrants").mockReturnValue("processGrants return value");
        });

        it("calls isSiret", () => {
            expect(vi.mocked(isSiret)).toHaveBeenCalledWith(IDENTIFIER);
        });

        it("fetches establishment grants", () => {
            expect(establishmentPort.getGrants).toHaveBeenCalledWith(IDENTIFIER);
        });

        it("fetches association grants", () => {
            vi.clearAllMocks();
            vi.mocked(isSiret).mockReturnValueOnce(false);
            CTRL = new GrantDashboardController(IDENTIFIER);
            // @ts-expect-error: mock
            vi.spyOn(CTRL, "processGrants").mockReturnValue("processGrants return value");
            expect(associationPort.getGrants).toHaveBeenCalledWith(IDENTIFIER);
        });

        it("calls processGrants", () => {
            // @ts-expect-error: access private method
            expect(CTRL.processGrants).toHaveBeenCalledWith(FLAT_GRANTS());
        });
    });

    describe("with a mocked constructor", () => {
        beforeAll(() => {
            vi.mocked(isSiret).mockReturnValueOnce(true);
            vi.mocked(establishmentPort.getGrants).mockResolvedValueOnce([]);
            CTRL = new GrantDashboardController(IDENTIFIER);
        });

        describe("initStores", () => {
            let updateRowSpy;
            beforeAll(() => {
                // @ts-expect-error -- test private
                updateRowSpy = vi.spyOn(CTRL, "updateRows").mockReturnValue();
            });

            afterAll(() => {
                updateRowSpy.mockRestore();
            });

            describe("selectedExercise derives from selectedExerciseIndex", () => {
                let updateSpy;
                let refStore;

                beforeAll(() => {
                    const derivedSpy = vi.spyOn(Store, "derived");
                    // @ts-expect-error -- test private
                    CTRL.initStores();
                    const callArgs = vi.mocked(derivedSpy).mock.calls[0];
                    updateSpy = callArgs[1];
                    refStore = callArgs[0];
                });

                it("depends on selectedExerciseIndex", () => {
                    const actual = refStore;
                    const expected = CTRL.selectedExerciseIndex;
                    expect(actual).toBe(expected);
                });

                it("returns null if no exercise index", () => {
                    const expected = null;
                    const actual = updateSpy(undefined);
                    expect(actual).toBe(expected);
                });

                it("returns null if no exerciseOptions", () => {
                    CTRL.exerciseOptions.value = undefined;
                    const expected = null;
                    const actual = updateSpy(0);
                    expect(actual).toBe(expected);
                });

                it("returns exercise from options", () => {
                    const expected = "2023";
                    CTRL.exerciseOptions.value = ["2022", expected];
                    const actual = updateSpy(1);
                    expect(actual).toBe(expected);
                });

                it("returns exercise from options with index set to 0", () => {
                    const expected = "2023";
                    CTRL.exerciseOptions.value = [expected, "2022"];
                    const actual = updateSpy(0);
                    expect(actual).toBe(expected);
                });
            });

            describe("on selectedExerciseIndex change", () => {
                let subscribeFn;
                const YEAR = "2021";
                const GRANTS = "g2021" as unknown as FlatGrant[];

                beforeAll(() => {
                    const subscribeSpy = vi.fn();
                    vi.spyOn(Store, "derived").mockReturnValueOnce({
                        subscribe: subscribeSpy,
                    } as unknown as ReadStore<string>);
                    // @ts-expect-error -- test private
                    CTRL.initStores();
                    subscribeFn = subscribeSpy.mock.calls[0][0];
                    CTRL.grantsByExercise = { [YEAR]: GRANTS };
                });

                it("does nothing if selectedExercise is set to null", () => {
                    const setSpy = vi.spyOn(CTRL.selectedGrants, "set");
                    subscribeFn(null);
                    expect(setSpy).not.toHaveBeenCalled();
                });

                it("updates selectedGrants through grantsByExercise on selectedExerciseIndex change", () => {
                    const setSpy = vi.spyOn(CTRL.selectedGrants, "set");
                    subscribeFn(YEAR);
                    expect(setSpy).toHaveBeenCalledWith(GRANTS);
                });
            });

            it("calls updateRows on selectedGrants change", () => {
                // @ts-expect-error -- test private
                CTRL.initStores();

                const GRANTS = "grants" as unknown as FlatGrant[];
                CTRL.selectedGrants.set(GRANTS);
                expect(updateRowSpy).toHaveBeenCalledWith(GRANTS);
            });
        });

        describe("processGrants", () => {
            const GRANTS_BY_EXERCISE = { 2021: GRANT_FROM_2021, 2023: [GRANT_FROM_2023], 2024: [GRANT_FROM_2024] };
            let splitSpy;
            beforeAll(() => {
                // @ts-expect-error: mock private method
                splitSpy = vi.spyOn(CTRL, "splitGrantsByExercise").mockReturnValue(GRANTS_BY_EXERCISE);
            });
            afterAll(() => {
                splitSpy.mockRestore();
            });

            it("defines grants", () => {
                // @ts-expect-error: calls private method
                CTRL.processGrants(FLAT_GRANTS());
                expect(CTRL.grants.value).toEqual(FLAT_GRANTS());
            });

            it("split grants by exercise", () => {
                // @ts-expect-error: calls private method
                CTRL.processGrants(FLAT_GRANTS());
                expect(CTRL.grantsByExercise).toEqual(GRANTS_BY_EXERCISE);
            });

            it("select most recent exercise", () => {
                // @ts-expect-error: calls private method
                CTRL.processGrants(FLAT_GRANTS());
                expect(CTRL.selectedExerciseIndex.value).toBe(2);
            });
        });

        describe("selectExercise", () => {
            it("update selectedExerciseIndex", () => {
                CTRL.selectExercise(1);
                expect(CTRL.selectedExerciseIndex.value).toBe(1);
            });
        });

        describe("clickProviderLink", () => {
            it("call trackerService", () => {
                CTRL.clickProviderLink();
                expect(vi.mocked(trackerService.trackEvent)).toHaveBeenCalledWith(
                    "association-etablissement.dashboard.display-provider-modal",
                );
            });
        });

        describe("splitGrantsByExercise", () => {
            it("returns grants by exercise", () => {
                // @ts-expect-error: call private method
                const actual = CTRL.splitGrantsByExercise(FLAT_GRANTS());
                expect(actual).toMatchSnapshot();
            });
        });

        describe("download", () => {
            beforeAll(() => {
                vi.mocked(establishmentService.getGrantExtract).mockImplementation(() =>
                    Promise.resolve({ blob: "blob", filename: "filename" }),
                );
                vi.mocked(associationService.getGrantExtract).mockImplementation(() =>
                    Promise.resolve({ blob: "blob", filename: "filename" }),
                );
            });

            it("stops if extract already loading", async () => {
                CTRL.isExtractLoading.set(true);
                await CTRL.download();
                expect(trackerService.trackEvent).not.toHaveBeenCalled();
            });

            it("calls trackerService if not already loading", async () => {
                CTRL.isExtractLoading.value = false;
                await CTRL.download();
                expect(trackerService.buttonClickEvent).toHaveBeenCalledWith(
                    "association-etablissement.dashbord.download-csv",
                    CTRL.identifier,
                );
            });

            it("set isExtractLoading to true", async () => {
                // @ts-expect-error: mock store
                CTRL.isExtractLoading = { set: vi.fn() };
                await CTRL.download();
                expect(CTRL.isExtractLoading.set).toHaveBeenNthCalledWith(1, true);
            });

            it("calls isSiret", async () => {
                vi.mocked(isSiret).mockReturnValue(true);
                await CTRL.download();
                expect(isSiret).toHaveBeenCalledWith(CTRL.identifier);
            });

            it("calls establishmentService", async () => {
                vi.mocked(isSiret).mockReturnValue(true);
                await CTRL.download();
                expect(establishmentService.getGrantExtract).toHaveBeenCalledWith(CTRL.identifier);
            });

            it("calls associationService", async () => {
                vi.mocked(isSiret).mockReturnValue(false);
                await CTRL.download();
                expect(associationService.getGrantExtract).toHaveBeenCalledWith(CTRL.identifier);
            });

            it("calls documentHelper.download()", async () => {
                await CTRL.download();
                expect(documentHelper.download).toHaveBeenCalled();
            });

            it("set isExtractLoading to false", async () => {
                // @ts-expect-error: mock store
                CTRL.isExtractLoading = { set: vi.fn() };
                await CTRL.download();
                expect(CTRL.isExtractLoading.set).toHaveBeenNthCalledWith(2, false);
            });
        });

        describe("updateRows", () => {
            let rowSetSpy;

            beforeAll(() => {
                rowSetSpy = vi.spyOn(CTRL.rows, "set");
            });

            it("does nothing if grants is null", () => {
                // @ts-expect-error -- test private
                CTRL.updateRows(null);
                expect(rowSetSpy).not.toHaveBeenCalled();
            });

            it("sets rows with results from adapters", () => {
                vi.mocked(isGranted).mockReturnValueOnce(true);
                vi.mocked(isGranted).mockReturnValueOnce(false);
                vi.mocked(getApplicationCells).mockImplementation(a => `ac-${a}` as unknown as TableCell[]);
                vi.mocked(getPaymentsCells).mockImplementation(a => `pc-${a}` as unknown as TableCell[]);
                vi.mocked(getApplicationDashboardData).mockImplementation(
                    a => `a-${a}` as unknown as DashboardApplication,
                );
                vi.mocked(getPaymentDashboardData).mockImplementation(a => `p-${a}` as unknown as DashboardPayment);
                const GRANTS = [
                    { application: "a0", payments: "p0" },
                    { application: "a1", payments: "p1" },
                ];
                // @ts-expect-error -- test private
                CTRL.updateRows(GRANTS);
                const actual = rowSetSpy.mock.calls[0][0];
                expect(actual).toMatchInlineSnapshot(`
                  [
                    {
                      "application": "a-a0",
                      "applicationCells": "ac-a0",
                      "granted": true,
                      "payment": "p-p0",
                      "paymentsCells": "pc-p0",
                    },
                    {
                      "application": "a-a1",
                      "applicationCells": "ac-a1",
                      "granted": false,
                      "payment": "p-p1",
                      "paymentsCells": "pc-p1",
                    },
                  ]
                `);
                vi.mocked(getApplicationCells).mockRestore();
                vi.mocked(getPaymentsCells).mockRestore();
                vi.mocked(getApplicationDashboardData).mockRestore();
                vi.mocked(getPaymentDashboardData).mockRestore();
            });
        });

        describe("sortTable", () => {
            it("changes sort order", () => {
                const expected = 2;
                // @ts-expect-error -- test mock private
                CTRL.columnsSortOrder[INDEX] = -2;
                CTRL.sortTable(INDEX);
                // @ts-expect-error -- test mock private
                const actual = CTRL.columnsSortOrder[INDEX];
                expect(actual).toBe(expected);
            });

            it("sorts row with compare function with proper index", () => {
                CTRL.sortTable(INDEX);
                expect(grantCompareFn[INDEX]).toHaveBeenCalled();
            });

            it("updates rows", () => {
                const spy = vi.spyOn(CTRL.rows, "update");
                CTRL.sortTable(INDEX);
                expect(spy).toHaveBeenCalled();
            });
        });

        describe.each`
            side            | method                  | keyWordInTracker | modalComponent
            ${"payment"}    | ${"onPaymentClick"}     | ${"payment"}     | ${PaymentsInfoModal}
            ${"subvention"} | ${"onApplicationClick"} | ${"subvention"}  | ${ApplicationInfoModal}
        `("onRowClick: $side side", ({ method, keyWordInTracker, modalComponent }) => {
            const INDEX = 1;
            const ROWS = [
                { paymentCells: "Pc0", applicationCells: "Ac0" },
                {
                    paymentCells: "Pc1",
                    applicationCells: "Ac1",
                },
            ] as unknown as SortableRow[];
            const SUBV = [
                { application: "A0", payments: "P0" },
                { application: "A1", payments: "P1" },
            ] as unknown as FlatGrant[];

            beforeAll(() => {
                CTRL.selectedGrants.value = SUBV;
                CTRL.rows.value = ROWS;
            });

            it("if no cells, nothing happens", () => {
                CTRL.rows.value = [ROWS[0], {} as SortableRow];
                CTRL[method](INDEX);
                expect(trackerService.buttonClickEvent).not.toHaveBeenCalled();
                expect(data.set).not.toHaveBeenCalled();
                expect(modal.set).not.toHaveBeenCalled();
                CTRL.rows.value = ROWS;
            });

            it("tracks button click", () => {
                CTRL[method](INDEX);
                const eventTag = `association-etablissement.dashbord.${keyWordInTracker}.more_information`;
                expect(trackerService.buttonClickEvent).toHaveBeenCalledWith(eventTag, IDENTIFIER);
            });

            it("sets data", () => {
                CTRL.selectedGrants.value = SUBV;
                CTRL[method](INDEX);
                expect(data.set).toHaveBeenCalledWith({ application: "A1" });
            });

            it("sets modal component", () => {
                CTRL.selectedGrants.value = SUBV;
                CTRL[method](INDEX);
                expect(modal.set).toHaveBeenCalledWith(modalComponent);
            });
        });
    });
});
