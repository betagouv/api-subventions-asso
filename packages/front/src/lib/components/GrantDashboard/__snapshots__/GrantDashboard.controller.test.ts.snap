// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GrantDashboard Controller > with a mocked constructor > splitGrantsByExercise > returns grants by exercise 1`] = `
{
  "2022": [
    {
      "application": {
        "annee_demande": 2022,
      },
      "payments": [
        {
          "dateOperation": "2023-01-13",
        },
        {
          "dateOperation": "2023-04-21",
        },
      ],
    },
  ],
  "2024": [
    {
      "application": {
        "annee_demande": 2024,
      },
    },
  ],
}
`;
